resilience4j-spring/src/main/java/io/github/resilience4j/circuitbreaker/configure/CircuitBreakerAnnotationConfigScanner.java

```diff
@@ -0,0 +1,135 @@
+/*
+ * Copyright 2019 Olov Andersson
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.github.resilience4j.circuitbreaker.configure;
+
+import java.lang.reflect.Method;
+import java.util.HashMap;
+import java.util.Map;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.BeansException;
+import org.springframework.beans.factory.config.BeanDefinition;
+import org.springframework.beans.factory.config.BeanFactoryPostProcessor;
+import org.springframework.beans.factory.config.ConfigurableListableBeanFactory;
+
+import io.github.resilience4j.circuitbreaker.annotation.CircuitBreaker;
+import io.github.resilience4j.common.circuitbreaker.configuration.CircuitBreakerConfigurationProperties.InstanceProperties;
+
+/**
+ * Scans all bean definitions for CircuitBreaker annotations and collects
+ * annotated configuration so it can be merged with the properties
+ * configuration. This is done as a BeanFactoryPostProcessor so we can ensure
+ * that we have found these before actual bean instantiation begins.
+ * <p>
+ * Unless resilience4j-spring-boot/resilience4j-spring-boot2/
+ * resilience4j-spring-cloud/resilience4j-spring-cloud2 is used this class
+ * must be explicitly added to the spring context as a bean for CircuitBreaker
+ * annotation exception config to be considered.
+ */
+public class CircuitBreakerAnnotationConfigScanner implements BeanFactoryPostProcessor {
```

Comment:
Should we ask the Spring Boot guys for an advice which is the simplest method to merge Annotation properties and ConfigurationProperties?

Reply:
I tried using a BeanPostProcessor first, but the problem with doing that is to ensure that the beans get initialized in the correct order. By using a BeanFactoryPostProcessor, we can be certain that we have picked up all the annotation configuration before the CircuitBreakerRegistry is created.

The CircuitBreakerConfigAnnotationScanner would need to postprocess all beans that might declare the annotation, but any bean that requires the CircuitBreakerRegistry must not get it injected before the scanner both has found all the annotations. So for instance a bean with the annotation would not be allowed to have the registry injected into itself, which would be a very hard restriction.

In particular, it would be very fragile for a library like this to end up with scenario that sets complicated requirements on the bean initialization order.

Reply:
...and the thing is, there's always that chance that you request a CircuitBreaker instance from the registry during bean injection, when you application starts up. So I think most other solutions would be hard to solve without setting restrictions on how you were allowed to work with the CircuitBreakerRegistry during application startup.

Reply:
@RobWin Tested the code with spring-cloud-config as you described above. It works as expected. The CircuitBreakerServiceRegistry bean is recreated when it is required after a refresh, and the configuration will be re-initialized the same way as before. Any config from annotations will merged as expected.